### PR TITLE
Resolve NLv2 incompatibility with multithreading

### LINKED ISSUE #3329: Error running Pyomo with threading
## Summary

I am trying to run several models in parallel using python's builtin `threading` module. However, when trying to solve, the `nl_writer` (`repn>plugins>nl_writer.py`) does not appear to be able to handle writing files in parallel. In particular, it uses a shared resource  `AMPLRepn.ActiveVisitor`, and a locking system which blocks any sort of parallel access from happening:

```python
def __enter__(self):
    assert AMPLRepn.ActiveVisitor is None
    AMPLRepn.ActiveVisitor = self.visitor
    self.pause_gc = PauseGC()
    self.pause_gc.__enter__()
    return self

def __exit__(self, exc_type, exc_value, tb):
    self.pause_gc.__exit__(exc_type, exc_value, tb)
    assert AMPLRepn.ActiveVisitor is self.visitor
    AMPLRepn.ActiveVisitor = None
```
If two threads attempt to write separate problem files at the same time, the first thread will set `ActiveVisitor`, while the other will fail at the `assert` statement, since `ActiveVisitor` was set by the first (since it is shared among threads). 

Note that this problem won't necessarily show up all of the time, since it relies on two threads trying to write `.nl` files at the same time (can be a rare occurrence) - it's more likely to happen with many threads and/or large problem files (?)

I am not sure if there is any easy way around this - the way this is implemented does appear to be an intentional design decision. 

### Steps to reproduce the issue

Below is an example python script which runs into the error pretty much all the time on my machine.

```python
import pyomo.environ as pyo
import threading


def thread_function():
    model = pyo.ConcreteModel()
    # create a lot of variables and bind them to constraints
    # (large model - will spend a lot time trying to write the problem file)
    # ie. x0 = 1, x1 = 1, x2 = 1, ...
    for i in range(200):
        model.add_component(f'x{i}', pyo.Var(domain=pyo.NonNegativeReals))
        model.add_component(f'c{i}', pyo.Constraint(expr=model.component(f'x{i}') == 1))

    # solve the model
    solver = pyo.SolverFactory('ipopt')
    solver.solve(model)


if __name__ == '__main__':
    threads = list()
    # create 500 threads to get a greater chance of hitting the error
    for i in range(500):
        x = threading.Thread(target=thread_function)
        threads.append(x)
        print(f"Thread {i} Start")
        x.start()

    for index, thread in enumerate(threads):
        thread.join()
        print(f"Thread {index} Done")
    
    print("Main Done")
```

### Error Message

```bash
line 16, in thread_function
    solver.solve(model)
  File "/home/alma/.local/lib/python3.10/site-packages/pyomo/opt/base/solvers.py", line 597, in solve
    self._presolve(*args, **kwds)
  File "/home/alma/.local/lib/python3.10/site-packages/pyomo/opt/solver/shellcmd.py", line 224, in _presolve
    OptSolver._presolve(self, *args, **kwds)
  File "/home/alma/.local/lib/python3.10/site-packages/pyomo/opt/base/solvers.py", line 706, in _presolve
    ) = self._convert_problem(
  File "/home/alma/.local/lib/python3.10/site-packages/pyomo/opt/base/solvers.py", line 757, in _convert_problem
    return convert_problem(
  File "/home/alma/.local/lib/python3.10/site-packages/pyomo/opt/base/convert.py", line 99, in convert_problem
    problem_files, symbol_map = converter.apply(*tmp, **tmpkw)
  File "/home/alma/.local/lib/python3.10/site-packages/pyomo/solvers/plugins/converter/model.py", line 184, in apply
    (problem_filename, symbol_map_id) = instance.write(
  File "/home/alma/.local/lib/python3.10/site-packages/pyomo/core/base/block.py", line 1991, in write
    (filename, smap) = problem_writer(self, filename, solver_capability, io_options)
  File "/home/alma/.local/lib/python3.10/site-packages/pyomo/repn/plugins/nl_writer.py", line 348, in __call__
    info = self.write(model, FILE, ROWFILE, COLFILE, config=config)
  File "/home/alma/.local/lib/python3.10/site-packages/pyomo/repn/plugins/nl_writer.py", line 391, in write
    with _NLWriter_impl(ostream, rowstream, colstream, config) as impl:
  File "/home/alma/.local/lib/python3.10/site-packages/pyomo/repn/plugins/nl_writer.py", line 548, in __enter__
    assert AMPLRepn.ActiveVisitor is None
AssertionError
```

### Information on your system

Pyomo version: 6.7.0
Python version: 3.11.7
Operating system: Windows (running WSL)
How Pyomo was installed: pip from within conda environment

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
The NLv2 writer makes use of a visitor to both walk / compile the model expressions and to generate the string representation of the nonlinear portion of expressions.  To accommodate this, the `AMPLRepn` result needed a handle on the visitor to be able to know what the current template is (terse or annotated).  We implemented this using a module-level singleton, which works for general serial Python processes, but is not thread-safe.  This PR redesigns the AMPLRepn so that the template belongs to the repn class and not the visitor, thereby removing the need for the singleton.

As a bonus, removing the singleton allows us to more easily break the visitor out of the writer module and into a separate `pyono.repn.ampl` module (following the pattern of the LP writer)

## Changes proposed in this PR:
- remove the reliance on a module-level singleton for communicating the active template from the writer/visitor to the AMPLRepn class
- split the AMPLRepn and AMPLRepnVisitor into a separate module

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/repn/ampl.py [hunk 1]
```diff
@@ -0,0 +1,1261 @@
...
+        True: vars_['less_than'],
+        False: vars_['less_equal'],
+        (True, True): (vars_['less_than'], vars_['less_than']),
+        (True, False): (vars_['less_than'], vars_['less_equal']),
+        (False, True): (vars_['less_equal'], vars_['less_than']),
+        (False, False): (vars_['less_equal'], vars_['less_equal']),
+    }
+
+
+class text_nl_debug_template(object):
```
Comment (mrmundt) [inline pyomo/repn/ampl.py:78]: FYI - don't need `(object)`.

File: pyomo/repn/ampl.py [hunk 2]
```diff
@@ -0,0 +1,1261 @@
...
+
+
+# The "standard" text mode template is the debugging template with the
+# comments removed
+class text_nl_template(text_nl_debug_template):
+    _strip_template_comments(vars(), text_nl_debug_template)
+    _create_strict_inequality_map(vars())
+
+
+class NLFragment(object):
```
Comment (mrmundt) [inline pyomo/repn/ampl.py:180]: FYI - don't need `(object)`.

File: pyomo/repn/ampl.py [hunk 3]
```diff
@@ -0,0 +1,1261 @@
...
+    def __init__(self, repn, node):
+        self._repn = repn
+        self._node = node
+
+    @property
+    def name(self):
+        return 'nl(' + self._node.name + ')'
+
+
+class AMPLRepn(object):
```
Comment (mrmundt) [inline pyomo/repn/ampl.py:200]: FYI - don't need `(object)`.

File: pyomo/repn/ampl.py (hunk 1 above)
Comment (mrmundt) [reply pyomo/repn/ampl.py:78]: Also this class name violates our own standards. I get that this is a legacy thing, but we may want to find places where we violate our own rules at some point.

File: pyomo/repn/ampl.py [hunk 4]
```diff
@@ -0,0 +1,1261 @@
...
+            for i, l in enumerate(v_lines):
+                comment_start = l.find('\t#')
+                if comment_start >= 0:
+                    v_lines[i] = l[:comment_start]
+            vars_[k] = '\n'.join(v_lines)
+
+
+# The "standard" text mode template is the debugging template with the
+# comments removed
+class text_nl_template(text_nl_debug_template):
```
Comment (mrmundt) [inline pyomo/repn/ampl.py:175]: This class name violates our own standards. I get that this is a legacy thing, but we may want to find places where we violate our own rules at some point.

**Review by mrmundt — APPROVED** [review]
I still want you to remove the `object`s but I won't let it hold this up...

File: pyomo/repn/ampl.py (hunk 4 above)
Comment (jsiirola) [reply pyomo/repn/ampl.py]: fixed!

**Review by mrmundt — APPROVED** [review]

**Review by emma58 — APPROVED** [review]
One docstring request, but this looks good.

File: pyomo/repn/plugins/nl_writer.py [hunk 5]
```diff
@@ -107,16 +74,12 @@
 
 logger = logging.getLogger(__name__)
 
-# Feasibility tolerance for trivial (fixed) constraints
-TOL = 1e-8
+relocated_module_attribute('AMPLRepn', 'pyomo.repn.ampl.AMPLRepn', version='6.4.0.dev0')
```
Comment (blnicho) [inline pyomo/repn/plugins/nl_writer.py:77]: Was this set to an old version on purpose?
```suggestion
relocated_module_attribute('AMPLRepn', 'pyomo.repn.ampl.AMPLRepn', version='6.7.4.dev0')
```

File: pyomo/repn/ampl.py [hunk 6]
```diff
@@ -0,0 +1,1322 @@
...
+                    if other.nonlinear.__class__ is list:
+                        self.nonlinear.extend(other.nonlinear)
+                    else:
+                        self.nonlinear.append(other.nonlinear)
+        elif _type is _CONSTANT:
+            self.const += other[1]
+
+    def to_expr(self, var_map):
+        if self.nl is not None or self.nonlinear is not None:
+            # TODO: support converting general nonlinear expressiosn
```
Comment (blnicho) [inline pyomo/repn/ampl.py:482]:
```suggestion
            # TODO: support converting general nonlinear expressions
```

**Review by blnicho — APPROVED** [review]
I found one other typo but otherwise this looks fine

**Pull request merged**